### ISSUE TITLE
Different commit hashes from git subtree split

### ISSUE THREAD

**Issue by mrmark** [splitsh/lite#11]
In testing `splitsh-lite` I was comparing the resulting split with what `git subtree split` produces and noticed that the resulting hashes are different at the end of the split.  In the tests for `splitsh-lite`, I noticed that the hashes should be the same regardless of which tool you use.  Unsure how to provide useful information here as the repository I'm using isn't publicly available.  When I run `splitsh-lite` it gives this output:

    472 commits created, 103972 commits traversed, in 8.962s

But when I count the commits with `git rev-list --count` it is actually 466.  When comparing the logs of the two splits, the `git subtree split` has all 472 commits.  The logs are identical until the last 49 commits.

My guess is that maybe some of the commits resulted in no changes somehow, which prevents the commit from taking place?  In the end, maybe this isn't really a problem unless we (for whatever reason) switch to `git subtree split`?

Please let me know if I can provide any other information as I know this report is likely lacking.

**Comment by fabpot** [comment]
The shas should be the same, but that depends on the version of Git you are using. If it is old, use `--legacy` with splitsh to get the same shas. If Git is recent enough, well, I have no ideas. This algorithm is used for very large projects with thousands of commits and split repos have the exact same shas as with the Git command. Can you at least paste the command you are running with both tools?

**Comment by mrmark** [comment]
Thank you very much for your reply, here are the commands:
```
> git --version
git version 2.9.0

> splitsh-lite --scratch --prefix=mod/hsuforum/ --origin=origin/master
472 commits created, 103972 commits traversed, in 8.675s
53fbe06911295f50fa2cdc743653a014d49ed27e

> git subtree split --prefix=mod/hsuforum -b subtree-mod_hsuforum-master origin/master
Created branch 'subtree-mod_hsuforum-master'
5ae323e621ea9481cb54fd99167d6c986b87f2ec
```

I tried with the `--legacy` option and that produced the same 466 commit count but the entire log has different shas from `git subtree split`.  Unsure if `splitsh-lite` uses the current `$PATH` for git?  Because I have it installed in `/opt/local/bin/git` (2.9.0) but also in `/usr/bin/git` (2.7.4).

**Comment by fabpot** [comment]
Has the oldest commit the same sha for both commands? If yes, can you spot which commit starts to diverge?

**Comment by mrmark** [comment]
So ran `git log --oneline --topo-order` on each split and compared the output.  This is what I'm seeing (from latest to oldest):
* 32 commits with different shas
* 17 commits which consist of a mix of commits that are identical in both splits and commits that _only_ exist in the `git subtree split`.
* Rest of the log is identical.

When I investigated the commits that only appear in the `git subtree split`, it appears we did some very poor branch management where basically the same change was merged in twice, but as different commits.  So I'm _guessing_ that those 6 missing commits are being ignored because there are no actual changes and since they are ignored, it throws off the latest 32 commits.   This is looking like a problem unique to our repository and I'm not sure if you can or even want to try to address it.  I'm going to try to split a couple of other directories and see if they generate better results.

Sorry for the trouble and thanks for the replies.  I would 100% understand if you would like to close this as "won't fix".  Cheers and thanks!

**Comment by mrmark** [comment]
Managed to recreate it, if you spit the `lib` directory in this branch: https://github.com/mrmark/testsplit/tree/mergeBthenA-Rebased

```
> splitsh-lite --scratch --prefix=lib && git subtree split --prefix=lib
6 commits created, 7 commits traversed, in 5ms
9a28d0f555da1fd57eac0d3e1037cdd1ece47943
d3ac33e803dcec933d0aa753287dd463b5240755
```

If I run with `--debug`, this is what it shows for the last two commits:
```
2016/07/01 08:48:20 Processing commit: c709c6e5d19de306d14dd5619b9e403a470025fe
2016/07/01 08:48:20   parents: 7e80f1d4b54753c51eac9cb28bdcbf238bcaa6b8
2016/07/01 08:48:20   newparents: 36c878fdc478a918901a74e3cf7029f1d2ac758e
2016/07/01 08:48:20   tree is: e9277fe4ab241818644342b309125705c42c48e6
2016/07/01 08:48:20   copy commit "c709c6e5d19de306d14dd5619b9e403a470025fe" "e9277fe4ab241818644342b309125705c42c48e6" "36c878fdc478a918901a74e3cf7029f1d2ac758e"
2016/07/01 08:48:20   newrev is: e2eb8eb3dd130a1b74220fbe425e8082fef24194
2016/07/01 08:48:20 Processing commit: 99dbed0a0dffb6c25ebc6f6d67deddcd04c51aa1
2016/07/01 08:48:20   parents: 562e6f50fe0b10c882ff4f04ba4a9caf949bffdf c709c6e5d19de306d14dd5619b9e403a470025fe
2016/07/01 08:48:20   newparents: 9a28d0f555da1fd57eac0d3e1037cdd1ece47943 e2eb8eb3dd130a1b74220fbe425e8082fef24194
2016/07/01 08:48:20   tree is: 41f595095084a726a76c89d5952e80d84bc04c55
2016/07/01 08:48:20   newrev is: 9a28d0f555da1fd57eac0d3e1037cdd1ece47943
6 commits created, 7 commits traversed, in 6ms
9a28d0f555da1fd57eac0d3e1037cdd1ece47943
```

Looks like that last one didn't copy.  Again, I was only able to do this via bad branch management where I introduce two identical commits but with different shas due to rebasing and merging both in.  Thanks for your patience with this issue, I should have lead with this information but I wasn't sure what was happening at the start.

**Comment by mrmark** [comment]
I was digging around in the code and it _appears_ that the fix would be in `copyOrSkip` as it is finding an identical parent.  I was trying to figure out what is unique about this use case and that got me thinking that the other parent might be the key.  The other parent was copied but it is never included in the history because the merge commit also has an identical parent (matching trees).  Unsure if some sort of logic could be added where if one of the parents was copied but not included in the branch history, then it would ignore the other identical parent and copy the commit to ensure all copied commits are in the resulting branch history.  Sorry if my ramblings are not making any sense.

**Comment by fabpot** [comment]
Thanks for the detailed analysis. It totally makes sense. I don't have time right now to have a closer look but I will try to have a look at a fix in the next few days.

**Comment by fabpot** [comment]
So, after some investigation, it seems like before git/git@a039a79e9d00fed442f31ce69054e98e7dd1fc69, both splitsh and git subtree split behaves in the same way. But after that commit, and so on git 2.9.0, they fixed a bug making history different.

I will see if we can make the same behavior with splitsh via a flag.

**Comment by fabpot** [comment]
@mrmark #12 should fix this issue. Can you confirm? If yes, I will add some more tests and a new flag for this new behavior.

**Comment by fabpot** [comment]
#12 is finished now, I've added some more tests based on your demo repo and a new flag to determine which behavior people want.

**Comment by mrmark** [comment]
Hooray! /happyDance

Confirmed in our repository that two splits are now generating the same shas as `git subtree split` when before they were not.  Thank you very much for your work on this!

I have to say, this is an amazing tool.  When testing, I ran `splitsh-lite` and it reports back in less than 10 seconds that it is done.  Then I run `git subtree split` and I go vacuum, do laundry, take out trash, then see it is done.
